src/server/pfs/server/driver.go

```diff
@@ -806,188 +700,116 @@ func (d *driver) repoSize(ctx context.Context, repo *pfs.Repo) (int64, error) {
 // commits arrive on 'branch', when 'branches's HEAD is deleted, or when
 // 'branches' are newly created (i.e. in CreatePipeline).
 func (d *driver) propagateBranches(txnCtx *txncontext.TransactionContext, branches []*pfs.Branch) error {
-	branchInfoCache := map[string]*pfs.BranchInfo{}
-	getBranchInfo := func(branch *pfs.Branch) (*pfs.BranchInfo, error) {
-		if branchInfo, ok := branchInfoCache[pfsdb.BranchKey(branch)]; ok {
-			return branchInfo, nil
-		}
-		branchInfo := &pfs.BranchInfo{}
-		if err := d.branches.ReadWrite(txnCtx.SqlTx).Get(branch, branchInfo); err != nil {
-			return nil, errors.EnsureStack(err)
-		}
-		branchInfoCache[pfsdb.BranchKey(branch)] = branchInfo
-		return branchInfo, nil
+	if len(branches) == 0 {
+		return nil
 	}
-
-	// subvBIMap = ( ⋃{b.subvenance | b ∈ branches} ) ∪ branches
-	// (after branches has been pruned of no-op members, which don't actually need propagation)
-	subvBIMap := map[string]*pfs.BranchInfo{}
-	for _, branch := range branches {
-		branchInfo, err := getBranchInfo(branch)
-		if err != nil {
-			return err
-		}
-
-		// We need to create new commits or aliases if any of this branch and its
-		// provenances disagree on their commit set.
-		ids := []string{branchInfo.Head.ID}
-		for _, provBranch := range branchInfo.Provenance {
-			provInfo, err := getBranchInfo(provBranch)
-			if err != nil {
-				return err
-			}
-			ids = append(ids, provInfo.Head.ID)
-		}
-		if allSameString(ids) && ids[0] != txnCtx.CommitSetID {
-			// this branch hasn't changed during the transaction, and is still consistent with its provenance,
-			// so we don't need to include it or its subvenance
-			continue
+	var propagatedBranches []*pfs.BranchInfo
+	seen := make(map[string]*pfs.BranchInfo, 0)
+	bi := &pfs.BranchInfo{}
+	for _, b := range branches {
+		if err := d.branches.ReadWrite(txnCtx.SqlTx).Get(b, bi); err != nil {
+			return errors.EnsureStack(err)
 		}
-
-		subvBIMap[pfsdb.BranchKey(branch)] = branchInfo
-		for _, subvBranch := range branchInfo.Subvenance {
-			subvInfo, err := getBranchInfo(subvBranch)
-			if err != nil {
-				return err
+		for _, sb := range bi.Subvenance {
+			if _, ok := seen[pfsdb.BranchKey(sb)]; !ok {
+				sbi := &pfs.BranchInfo{}
+				if err := d.branches.ReadWrite(txnCtx.SqlTx).Get(sb, sbi); err != nil {
+					return errors.Wrapf(err, "get subvenant branch %q", pfsdb.BranchKey(sb))
+				}
+				seen[pfsdb.BranchKey(sb)] = sbi
+				propagatedBranches = append(propagatedBranches, seen[pfsdb.BranchKey(sb)])
```

Comment:
nit, I would just use the branch info variable here rather than the map lookup. Also, for a scope this small, `b` and `bi` seem slightly nicer.

Reply:
for sure 👍